content/docs/testing-recipes.md

```diff
@@ -0,0 +1,403 @@
+        or on their <a href={props.site}>website</a>.
+      </address>
+      <Map center={props.center} />
+    </div>
+  );
+}
+```
+
+In a testing environment, it's not very useful to load the Map component, besides also being inessential to testing our own code. In this situation, we can mock out the dependency itself to a dummy component, and run our tests.
+
+```jsx
+// contact.test.js
+import React from "react";
+import { render } from "react-dom";
+import Contact from "./contact";
+import MockedMap from "./map";
+
+jest.mock("./map", () => jest.fn(() => "DummyMap"));
+
+let container = null
+beforeEach(() => {
+  // setup a DOM element as a render target
+  container = document.createElement('div')
+  document.body.appendChild(container)
+})
+
+afterEach(() => {
+  // cleanup on exiting
+  container.remove();
+  container = null  
+})
+
+it("should render contact information", () => {
+  const center = { lat: 0, lang: 0 };
+  act(() => {
+    render(<Contact name="" email="" site="" center={center} />, container);
+  });
+  // ensure the mocked map component function was called correctly
+  expect(MockedMap).toHaveBeenCalledWith({ id: "example-map", center }, {});
+});
+```
+
+### Events {#events}
+
+Dispatch real events on components and elements, and observe the side effects of those actions on UI elements. Consider a `Toggle` component:
+
+```jsx
+// toggle.js
+import React, { useState } from "react";
+function Toggle(props) {
+  // initial, onChange
+  const [state, setState] = useState(Boolean(initial));
+  return (
+    <button
+      onClick={() => {
+        setState(previousState => !previousState);
+        onChange(!previousState);
+      }}
+      data-testid="toggle"
+    >
+      {state === true ? "Turn off" : "Turn on"}
+    </button>
+  );
+}
+```
+
+We could write tests for it:
+
+```jsx
+// toggle.test.js
+import React from "react";
+import { render } from "react-dom";
+import Toggle from "./toggle";
+
+let container = null
+beforeEach(() => {
+  // setup a DOM element as a render target
+  container = document.createElement('div')
+  document.body.appendChild(container)
+})
+
+afterEach(() => {
+  // cleanup on exiting
+  container.remove();
+  container = null  
+})
+
+it("changes value when clicked", () => {
+  // a counter to track clicks
+  let clicks = 0;
+  act(() => {
+    render(<Toggle initial={true} onChange={() => clicks++} />, container);
+  });
+  // get a hold of the button element, and trigger some clicks on it
+  const button = document.querySelector("[data-testid=toggle]");
+  expect(button.innerText).toBe("Turn on!");
+  act(() => {
+    button.dispatchEvent(new MouseEvent("click", { bubbles: true }));
+  });
+  expect(clicks).toBe(1);
+  expect(button.innerText).toBe("Turn off!");
+  act(() => {
+    for (let i = 0; i < 5; i++) {
+      button.dispatchEvent(new MouseEvent("click", { bubbles: true }));
+    }
+  });
+  expect(clicks).toBe(6);
+  expect(button.innerText).toBe("Turn on!");
+});
+```
```

Comment:
Call out here that it *needs to be attached to document* for events to fire. This is easy to miss.